CSS-like Animations & Transitions

jmacato opened the pull request:
This PR tracks our attempt at implementing a CSS-like Style-definable Keyframe Animations and Property Transitions for Avalonia.
 
### Non-Target
- Extensive unit-tests (that will be filed after this PR).
- Alternative KeyFrame interpolation algorithms (Splines, Bezier, etc.)

## To-do list

#### Property Transitions
- [x] Style-declared Transitions 
- [x] Control-declared Transitions
- [x] Numerical (`Double`, `Float`, `Integer`) Transitions
 ~~`SolidColorBrush` Transition~~ Will be on a future PR.
#### Animations
- [x] Style-declared Animations
- [x] Proper Class Triggering behavior
- [x] Global Animations Timer with controllable play states 
- [x] Key frames
    - [x] Generic `KeyFrame` objects
    - [x] Numerical (`Double`, `Float`, `Integer`) Keyframes Animations
    ~~Color (`SolidColorBrush`) Keyframes Animation~~ Will be on a future PR.
- [x] Controllable Play state on every `Animatable`-derived controls.
- [x] Repeat Behavior
- [x] Fill modes (similar to [this](https://www.w3schools.com/cssref/css3_pr_animation-fill-mode.asp))

#### Miscellaneous 
- [ ] Rework PageTransitions to use the new animations backend.
- [ ] Rework ProgressBar's indeterminate state animation.

Fixes #778

Commit:
Rename Carousel & TabControls Transition property to PageTransitions to avoid confusion to the upcoming Transitions from Avalonia.Animations.

Commit:
*Added initial implementation of new Transitions
*Added default easing classes
*Added easing type converter

Commit:
*Added initial implementation of new Transitions
*Added default easing classes
*Added easing type converter

Commit:
Merge conflicting commits

Commit:
*Commented out stray xml comment.

Commit:
*Commented out stray xml comment.

Commit:
Merge branch 'wip-animations' of https://github.com/jmacato/Avalonia into wip-animations

Commit:
Fixed transition backend, transitions are now working.

Commit:
*Added numerical transition classes.
*Disabled old animations code in CrossFade, PageSlide, CarouselPresenter and ProgressBar.

Commit:
Revert the library split, just transferring the base Animatable class to Avalonia.Visual was good enough.
Apply important changes from my internal test branch

grokys committed:
Made visuals depend on animations.

Instead of `Avalonia.Animations` depending on `Avalonia.Visuals`, make it the other way round.

Commit:
*Implement Basic Keyframe Animations support.
*Implement DoubleKeyFrames for Properties such as Opacity, etc.
*Implement TransformKeyFrames with initial implementation
 of specialized logic for selecting RenderTransform of the
 target control properly.
*Ported RenderTest to .NET Core for testing and to remove the
 crufty old .csproj format.
*Replaced AnimationsPage with some samples of hover-activated
 animated red rectangles.

Commit:
* Fixed interpolation point selection when theres three or more keyframes.

*Added new examples in RenderTest/AnimationsPage

Commit:
Finally, finally fix the linear parametric interpolation for DoubleKeyFrames 😄

Credits to Georgia Tech's keyframe slides : https://www.cc.gatech.edu/classes/AY2012/cs4496_spring/slides/Keyframe.pdf

Commit:
* Transfer some keyframe handling responsibilites to the base class for maximum reusability on later custom derived classes.

* Fix flickering issues on restart in TransformKeyFrames.

*Add some syntactic sugar here and there.

jmacato commented:
Progress / Changes Report
- Property Transitions are now functionally complete, thanks to prior work on them by @grokys and others.
- Global animation clock (might add some extensions to control this timer globally but it's for another time).
- Added Easing Classes that are derived from this (effectively) public domain code: https://github.com/acron0/Easings
- Automatically detect all Easing Classes and adds them to a static dictionary, enabling a simpler selection of desired easings. (only runs once for performance reasons).
- Implemented Animations and KeyFrames.
- Implemented special logic for `TransformKeyFrames` in selecting the necessary Transform object in the target controls.
- Animations are defined in Styles for now since they serve as the WPF's equivalent of `Triggers`.

To Do
- Implement Play-Pause-Stop functions to specific controls or globally (if design allows) 
- Polish the code and xml comments a bit more.

Non-Target
- Extensive unit-tests (this will be filed after this PR).

Commit:
Minor fixes.

Commit:
Minor Fixes pt. 2

Commit:
Pass target control to the keyframes and other animation classes.
Make the global timer discrete to avoid costly TimeSpan conversions and enable global animation pause/play support.

Commit:
Split the global timer for Transitions and Animations.
Added more example animations in RenderTest with Global Pause/Resume button for Animations.

Commit:
Fix transitions

Commit:
TransitionKeyFrames transform selection logic are now fully functional.

Trigger RaiseChanged() on TransformGroup class when a child transform was changed.

Split the PlayState enum to a new file.

Commit:
Minor fix.

Commit:
Sync with master.

Commit:
Some tweaking with AnimationPage.xaml.
Merged RenderTest's Program and App class for brevity.

Commit:
fix mismatched end tag in ControlCatalog/MainView.xaml

Commit:
- Initial implementation of ColorTransition.cs
- Added System.Numerics.Vectors in Avalonia.Base.
- Replaced AnimationPlayState in Animatable.cs with a DirectProperty.

Commit:
Forgot to include Base.props in the last commit.

Commit:
Fix color transitions code. Reworked animation example page a little bit.

Commit:
Fix formatting in AnimationsPage.xaml

Commit:
Keeping it simple, ColorTransition will now just use sRGB premultiplied space
color interpolation.

Commit:
Removed ColorTransitions. Will be revived in future PR.

Commit:
Removed ColorTransitions from AnimationsPage.xaml

Commit:
Added RepeatBehavior, Direction and PlayState
enums and their respective properties in Animatable.cs

Commit:
Remove the 2nd Observable.Tick and merge the transitions and animations timer.

Commit:
Fix a nit

Commit:
Delegated the animation clock to the Animatable objects so that we can control their Play State individually.

Commit:
Implement IterationTokens for Repeat behaviors

Commit:
Initial Implementation of KeyFrame state machines for handling different stages of an animation.

Commit:
Amended previous commit.

Commit:
Addendum to the previous commit, again.

Commit:
Make KeyFrameStateMachine an IObservable.

Commit:
Fixed delay logic on keyframe statemachine.

Commit:
make KFSM.Step a void, doesnt make sense to make it return some value now.

Commit:
removed spurrious code

Commit:
Commented out System.Numerics.Vectors and the corresponding Vector4 to Color conversion code.

Completed the Repeat Behaviors logic (Loop, Repeat)

State Machine can now respond to PlayStates properly.

Combined the Intra-frame normalized time and Keyframe Pair selection into a single function for flexibility and customizable custom KeyFrames objects on the end-user.

Initial FillMode implementation.

Commit:
Implemented FillMode; This PR is now almost feature complete (that nasty Control.Class selector bug).

Commit:
Minor Fixes on KFSM regarding backward fill mode.
Reimplemented ClippingPage.xaml to XAML Keyframes.
Reverted Animatable properties to a DirectProperty.

Commit:
Revert changes in AppBuilder.cs

Commit:
Fix tests and rename classes with typeparams according to conventions.

Commit:
Merge branch 'master' of https://github.com/AvaloniaUI/Avalonia into wip-animations

Commit:
Added a looping example. Removed Easing from default type converters, adding an typeconverter attribute is enough.

Commit:
fixed message in Transition`1.cs

Commit:
Merge branch 'master' of https://github.com/AvaloniaUI/Avalonia into wip-animations

Commit:
Merge branch 'master' of https://github.com/AvaloniaUI/Avalonia into wip-animations

grokys commented on src/Avalonia.Animation/Keyframes/KeyFrame.cs:
```diff
@@ -0,0 +1,70 @@
+        /// Gets or sets the cue of this <see cref="KeyFrame"/>.
+        /// </summary>
+        /// <value>The cue.</value>
+        public Cue Cue
```
So a `Cue` is a time as a percent of the animation and a `KeyTime` is an actual time?

jmacato replied:
Yes, i added the `KeyTime` as an alternative when they want to exactly pinpoint when in the animation iteration a keyframe should be activated and also they can't set both `Cue` and `KeyTime` at the same time :D

grokys commented on src/Avalonia.Animation/Keyframes/KeyFrame.cs:
```diff
@@ -0,0 +1,70 @@
+        public object Value { get; set; }
+
+    }
+
```
whitespace :)

grokys commented on src/Avalonia.Animation/Keyframes/KeyFrames`1.cs:
```diff
@@ -0,0 +1,186 @@
+        /// </summary>
+        public Dictionary<double, T> ConvertedKeyframes = new Dictionary<double, T>();
+
+        private bool IsVerfifiedAndConverted;
```
Could we put fields at the top? Also it's capitalized like a property.

grokys commented on src/Avalonia.Animation/Keyframes/IKeyFrames.cs:
```diff
@@ -0,0 +1,17 @@
+    /// <summary>
+    /// Interface for Keyframe group object
+    /// </summary>
+    public interface IKeyFrames
```
I'm interested to know why someone would want to provide a different implementation of `IKeyFrames`?

grokys commented on src/Avalonia.Animation/Keyframes/KeyFrames`1.cs:
```diff
@@ -0,0 +1,186 @@
+        public abstract T DoInterpolation(double time);
+
+
+        // public abstract IObservable<T> DoInterpolation(IObservable<double> timer, Animation animation,
```
Commented out code.

grokys commented on src/Avalonia.Animation/Keyframes/KeyFrames`1.cs:
```diff
@@ -0,0 +1,186 @@
+            hasStartKey = hasEndKey = false;
+
+            // this can be optional later, by making the default start/end keyframes
+            // to have a neutral value (a.k.a. the value prior to the animation).
```
What does this comment mean? Is it a TODO or is it explaining the logic here?

grokys commented on src/Avalonia.Animation/Keyframes/KeyFrames`1.cs:
```diff
@@ -0,0 +1,186 @@
+        /// <inheritdoc/>
+        public virtual IDisposable Apply(Animation animation, Animatable control, IObservable<bool> obsMatch)
+        {
+            if (!IsVerfifiedAndConverted)
```
Do we want to invalidate this flag if keyframes are added/removed after `VerifyConvertKeyFrames` is called? Or should we just not support that?

grokys commented on src/Avalonia.Animation/Keyframes/KeyFrames`1.cs:
```diff
@@ -0,0 +1,186 @@
+    {
+
+        /// <summary>
+        /// Target property.
```
"Gets or sets the target property for the keyframe."

grokys commented on src/Avalonia.Animation/Keyframes/KeyFrames`1.cs:
```diff
@@ -0,0 +1,186 @@
+        /// <summary>
+        /// List of type-converted keyframes.
+        /// </summary>
+        public Dictionary<double, T> ConvertedKeyframes = new Dictionary<double, T>();
```
Does this need to be public? It's only used by this class.

grokys commented on src/Avalonia.Animation/Keyframes/KeyFrames`1.cs:
```diff
@@ -0,0 +1,186 @@
+        /// time parameter).
+        /// </summary>
+        /// <param name="t">The time parameter, relative to the total animation time</param>
+        public (double IntraKFTime, KeyFramePair<T> KFPair) GetKFPairAndIntraKFTime(double t)
```
I think this can be make `protected`?

grokys commented on src/Avalonia.Animation/Keyframes/KeyFrames`1.cs:
```diff
@@ -0,0 +1,186 @@
+        /// <summary>
+        /// Runs the KeyFrames Animation.
+        /// </summary>
+        public IDisposable RunKeyFrames(Animation animation, Animatable control)
```
I think this can be `private`?

grokys commented on src/Avalonia.Animation/Keyframes/KeyFrames`1.cs:
```diff
@@ -0,0 +1,186 @@
+        /// <summary>
+        /// Interpolates a value given the desired time.
+        /// </summary>
+        public abstract T DoInterpolation(double time);
```
I think this can be `protected`?

grokys commented on src/Avalonia.Animation/Keyframes/KeyFramesStateMachine`1.cs:
```diff
@@ -0,0 +1,276 @@
+                    break;
+            }
+
+            switch (animation.PlaybackDirection)
```
I think this could just be:

```
_isReversed = animation.PlaybackDirection == PlaybackDirection.Reverse ||
                       animation.PlaybackDirection == PlaybackDirection.AlternateReverse;
```

grokys replied:
Or even make `PlaybackDirection` a `Flags` and do `_isReversed = (animation.PlaybackDirection & PlaybackDirection.Reverse) != 0;`

grokys commented on src/Avalonia.Animation/Keyframes/KeyFramesStateMachine`1.cs:
```diff
@@ -0,0 +1,276 @@
+
+        private enum KeyFramesStates
+        {
+            INITIALIZE,
```
I know these are private, but `ALL_CAPS` enums are sooo java ;)

jmacato replied:
Hahaha right sir, will fix this one. I was in the influence of C++ when i was writing this one :laughing:

grokys commented on src/Avalonia.Animation/Keyframes/KeyFramesStateMachine`1.cs:
```diff
@@ -0,0 +1,276 @@
+            _targetObserver = observer;
+            return this;
+        }
+        public void Dispose()
```
Missing newline

grokys commented on src/Avalonia.Animation/Keyframes/KeyFramesStateMachine`1.cs:
```diff
@@ -0,0 +1,276 @@
+                    if (_fillMode == FillMode.Forward
+                     || _fillMode == FillMode.Both)
+                    {
+                        _targetControl.SetValue(_parent.Property, _lastInterpValue, BindingPriority.Animation);
```
Hmm, should this be set with `Animtion` or `LocalValue` priority? I'm not sure...

grokys commented on src/Avalonia.Animation/Timing.cs:
```diff
@@ -0,0 +1,123 @@
+        /// </summary>
+        static Timing()
+        {
+            var globalTimer = Observable.Interval(FrameTick, AvaloniaScheduler.Instance);
```
Eventually we'll want this to be using the same timer as the renderer loop, but that will be a separate PR. Just leaving a reminder here :)

grokys commented on src/Avalonia.Styling/Styling/Style.cs:
```diff
@@ -78,6 +82,17 @@ public IResourceDictionary Resources
+            {
+                return _animations ?? (_animations = new List<IAnimation>());
+            }
+            set
```
Does this need to be settable?

grokys commented on src/Avalonia.Styling/Styling/Style.cs:
```diff
@@ -101,6 +116,22 @@ public void Attach(IStyleable control, IStyleHost container)
+                    foreach (var animation in Animations)
+                    {
+                        // TODO: Needs more work in passing the appropriate
+                        //       observable.
```
What more is needed here?

grokys commented on src/Avalonia.Visuals/Animation/Keyframes/TransformKeyFrames.cs:
```diff
@@ -0,0 +1,78 @@
+namespace Avalonia.Animation.Keyframes
+{
+    /// <summary>
+    /// Key frames that handles <see cref="double"/> properties.
```
Should this be "that handles Transform properties."? or "that handles double properties on Transforms"?

grokys commented on src/Avalonia.Visuals/Animation/Keyframes/TransformKeyFrames.cs:
```diff
@@ -0,0 +1,78 @@
+            }
+            else
+            {
+                throw new Exception($"Unsupported property {Property}");
```
Maybe be a bit more helpful here: `$"Cannot apply keyframe: property {Property} is not a Transform"` or something?

grokys commented on src/Avalonia.Visuals/Animation/Keyframes/TransformKeyFrames.cs:
```diff
@@ -0,0 +1,78 @@
+            }
+        }
+
+        void InitializeInternalDoubleKeyFrames()
```
`InitializeChildKeyFrames`?

grokys commented on src/Avalonia.Visuals/Animation/Keyframes/TransformKeyFrames.cs:
```diff
@@ -0,0 +1,78 @@
+        }
+
+        /// <inheritdocs/>
+        public override double DoInterpolation(double time) => 0;
```
Does this always return 0 because it's a TODO? Or can we not interpolate for transforms?

jmacato replied:
We can't directly interpolate `Transform` object, hence `TransformKeyFrames` is handling it in a roundabout way (proxying and applying the necessary logic to target the specified Transform type and Property)

grokys commented on src/Avalonia.Visuals/Media/Color.cs:
```diff
@@ -81,6 +82,32 @@ public static Color FromUInt32(uint value)
             );
         }
 
+        // /// <summary>
```
Commented-out code.

grokys commented on src/Avalonia.Visuals/Media/TransformGroup.cs:
```diff
@@ -17,6 +18,29 @@ public class TransformGroup : Transform
         public TransformGroup()
         {
             Children = new Transforms();
+            Children.CollectionChanged += (_, e) =>
```
You can use [`ForEachItem`](https://github.com/AvaloniaUI/Avalonia/blob/d7036e0afe1b674401756b8966124c24b6620d1e/src/Avalonia.Base/Collections/AvaloniaListExtensions.cs#L41) to do this a little more succinctly I think. Also make sure that `Transforms` sets its `ResetBehavior` to `Remove`.

grokys reviewed (commented):
Ok, I know this is still a WIP, but given it an initial review. Super cool stuff, this is really powerful and the syntax looks so much better than WPF's animation!

Commit:
Moved out KeyFrames and Transitions namespaces back
to Avalonia.Animations

KvanTTT commented on src/Avalonia.Animation/Easing/BackEaseIn.cs:
```diff
@@ -0,0 +1,21 @@
+        public override double Ease(double progress)
+        {
+            double p = progress;
+            return p * p * p - p * Math.Sin(p * Math.PI); 
```
Can be optimized to `p * (p * p - Math.Sin(p * Math.PI))`? Minus one multiplication.

KvanTTT commented on src/Avalonia.Animation/Easing/BackEaseInOut.cs:
```diff
@@ -0,0 +1,32 @@
+            if (p < 0.5d)
+            {
+                double f = 2d * p;
+                return 0.5d * (f * f * f - f * Math.Sin(f * Math.PI));
```
`0.5d * f * (f * f - Math.Sin(f * Math.PI))`

KvanTTT commented on src/Avalonia.Animation/Easing/BackEaseInOut.cs:
```diff
@@ -0,0 +1,32 @@
+            else
+            {
+                double f = (1d - (2d * p - 1d));
+                return 0.5d * (1d - (f * f * f - f * Math.Sin(f * Math.PI))) + 0.5d;
```
`0.5d * (1d - f * (f * f - Math.Sin(f * Math.PI))) + 0.5d`

KvanTTT commented on src/Avalonia.Animation/Easing/BackEaseOut.cs:
```diff
@@ -0,0 +1,21 @@
+        public override double Ease(double progress)
+        {
+            double p = 1d - progress;
+            return 1 - (p * p * p - p * Math.Sin(p * Math.PI));
```
`1 - p * (p * p - Math.Sin(p * Math.PI))`

KvanTTT commented on src/Avalonia.Animation/Easing/CircularEaseIn.cs:
```diff
@@ -0,0 +1,23 @@
+        public override double Ease(double progress)
+        {
+            double p = progress;
+            return 1d - Math.Sqrt(1d - (p * p));
```
Parenthesis are not required for multiplication: `1d - Math.Sqrt(1d - p * p)`.

KvanTTT commented on src/Avalonia.Animation/Easing/CircularEaseInOut.cs:
```diff
@@ -0,0 +1,29 @@
+            double p = progress;
+            if (p < 0.5d)
+            {
+                return 0.5d * (1d - Math.Sqrt(1d - 4d * (p * p)));
```
Useless parenthesis for multiplication.

KvanTTT commented on src/Avalonia.Animation/Easing/CircularEaseInOut.cs:
```diff
@@ -0,0 +1,29 @@
+            }
+            else
+            {
+                return 0.5d * (Math.Sqrt(-((2d * p) - 3d) * ((2d * p) - 1d)) + 1d);
```
Useless parenthesis for multiplication.

jmacato replied:
Can you please specify which parentheses to omit? also with the others too, if you may :)

KvanTTT replied:
`0.5d * (Math.Sqrt(-(2d * p - 3d) * (2d * p - 1d)) + 1d)`.

Moreover, it can be simplified to:

```csharp
double t = 2d * p;
return 0.5d * (Math.Sqrt(-(t - 3d) * (t - 1d)) + 1d);
```

jmacato replied:
thanks for these optimizations! i'll try to commit these later when i get back :smile:

KvanTTT commented on src/Avalonia.Animation/Easing/CubicEaseInOut.cs:
```diff
@@ -0,0 +1,29 @@
+            }
+            else
+            {
+                double f = ((2d * p) - 2d);
```
Useless parenthesis for multiplication.

KvanTTT commented on src/Avalonia.Animation/Easing/ElasticEaseInOut.cs:
```diff
@@ -0,0 +1,31 @@
+
+            if (p < 0.5d)
+            {
+                return 0.5d * Math.Sin(13d * EasingUtils.HALFPI * (2d * p)) * Math.Pow(2d, 10d * ((2d * p) - 1d));
```
Useless parenthesis for multiplication.

KvanTTT commented on src/Avalonia.Animation/Easing/ExponentialEaseInOut.cs:
```diff
@@ -0,0 +1,30 @@
+
+            if (p < 0.5d)
+            {
+                return 0.5d * Math.Pow(2d, (20d * p) - 10d);
```
Useless parenthesis for multiplication.

KvanTTT commented on src/Avalonia.Animation/Easing/ExponentialEaseInOut.cs:
```diff
@@ -0,0 +1,30 @@
+            }
+            else
+            {
+                return -0.5d * Math.Pow(2d, (-20d * p) + 10d) + 1d;
```
Useless parenthesis for multiplication.

KvanTTT commented on src/Avalonia.Animation/Easing/QuadraticEaseInOut.cs:
```diff
@@ -0,0 +1,28 @@
+            }
+            else
+            {
+                return (-2d * p * p) + (4d * p) - 1d;
```
Useless parenthesis for multiplication.

KvanTTT commented on src/Avalonia.Animation/Easing/QuarticEaseIn.cs:
```diff
@@ -0,0 +1,20 @@
+        /// <inheritdoc/>
+        public override double Ease(double progress)
+        {
+            return progress * progress * progress * progress;
```
Can be optimized to:

```csharp
double p2 = progress * progress;
return p2 * p2;
```

KvanTTT commented on src/Avalonia.Animation/Easing/QuarticEaseInOut.cs:
```diff
@@ -0,0 +1,29 @@
+            {
+                double f = (p - 1d);
+                return -8d * f * f * f * f + 1d;
+            }           
```
```csharp
if (p < 0.5d)
{
    double p2 = p * p;
    return 8d * p2 * p2;
}
else
{
    double f = p - 1d;
    double f2 = f * f;
    return -8d * f2 * f2 + 1d;
}
```

KvanTTT commented on src/Avalonia.Animation/Easing/QuarticEaseOut.cs:
```diff
@@ -0,0 +1,20 @@
+        public override double Ease(double progress)
+        {
+            double f = (progress - 1d);
+            return f * f * f * (1d - progress) + 1d;
```
```csharp
double f = progress - 1d;
double f2 = f * f;
return - f2 * f2 + 1d;
```

KvanTTT commented on src/Avalonia.Animation/Easing/QuinticEaseIn.cs:
```diff
@@ -0,0 +1,18 @@
+        /// <inheritdoc/>
+        public override double Ease(double progress)
+        {
+            return progress * progress * progress * progress * progress;
```
```csharp
double p2 = progress * progress;
return p2 * p2 * progress;
```

KvanTTT commented on src/Avalonia.Animation/Easing/QuinticEaseInOut.cs:
```diff
@@ -0,0 +1,28 @@
+            {
+                double f = ((2d * p) - 2d);
+                return 0.5d * f * f * f * f * f + 1d;
+            }
```
```csharp
if (progress < 0.5d)
{
    double p2 = p * p;
    return 16d * p2 * p2 * p;
}
else
{
    double f = (2d * p) - 2d;
    double f2 = f * f;
    return 0.5d * f2 * f2 * f + 1d;
}
```

KvanTTT commented on src/Avalonia.Animation/Easing/QuinticEaseOut.cs:
```diff
@@ -0,0 +1,19 @@
+        public override double Ease(double progress)
+        {
+            double f = (progress - 1d);
+            return f * f * f * f * f + 1d;
```
```csharp
double f2 = f * f;
return f2 * f2 * f + 1d;
```

KvanTTT reviewed (commented):
I added some math expression optimization notes.

Commit:
Addressed all review notes by sir @grokys .

Commit:
Fix protected overrides and restore IKeyFrame interface.

Commit:
Merge branch 'master' into wip-animations

Commit:
Optimized Easing Functions courtesy of @KvanTTT

jmacato commented:
Thanks @KvanTTT for your optimization notes!

Commit:
Merge branch 'master' into wip-animations

Commit:
Merge branch 'master' into wip-animations

Commit:
Merge branch 'master' into wip-animations

Commit:
Merge master

Commit:
Added class-selector bug sample in RenderTest

grokys committed:
Only trigger class match when value changed.

Previously each time a class was added to a control, the selector match would be re-triggered even if it didn't affect whether the selector matches the class.

Commit:
Merge branch 'master' into wip-animations

Commit:
Restore AnimationsPage.xaml

danwalmsley commented on src/Avalonia.Animation/KeyFramePair`1.cs:
```diff
@@ -0,0 +1,43 @@
+    /// </summary>
+    public struct KeyFramePair<T>
+    {
+
```
linespace

danwalmsley commented on src/Avalonia.Animation/KeyFramesStateMachine`1.cs:
```diff
@@ -0,0 +1,270 @@
+        object _lastInterpValue;
+        object _firstKFValue;
+
+        private ulong _delayTotalFrameCount,
```
can we declare each of these on a separate line, without , separation.

danwalmsley commented on src/Avalonia.Animation/KeyFramesStateMachine`1.cs:
```diff
@@ -0,0 +1,270 @@
+            _repeatCount,
+            _currentIteration;
+
+        private bool _isLooping,
```
separate declaration for each field.

danwalmsley commented on src/Avalonia.Animation/KeyFramesStateMachine`1.cs:
```diff
@@ -0,0 +1,270 @@
+                     _gotFirstKFValue;
+        private FillMode _fillMode;
+        private PlaybackDirection _animationDirection;
+        private KeyFramesStates _currentState, _savedState;
```
separate declaration.

danwalmsley commented on src/Avalonia.Animation/KeyFramesStateMachine`1.cs:
```diff
@@ -0,0 +1,270 @@
+                _currentState = KeyFramesStates.DoDelay;
+            else
+                _currentState = KeyFramesStates.DoRun;
+
```
linespace

danwalmsley commented on src/Avalonia.Animation/KeyFramesStateMachine`1.cs:
```diff
@@ -0,0 +1,270 @@
+
+        private void InternalStep(PlayState _playState, Func<double, T> Interpolator)
+        {
+
```
line space

danwalmsley commented on src/Avalonia.Animation/KeyFramesStateMachine`1.cs:
```diff
@@ -0,0 +1,270 @@
+                        goto checkstate;
+                    }
+                    _delayFrameCount++;
+
```
line space.

danwalmsley commented on src/Avalonia.Animation/KeyFramesStateMachine`1.cs:
```diff
@@ -0,0 +1,270 @@
+                    _targetObserver.OnCompleted();
+                    break;
+            }
+
```
ls

danwalmsley commented on src/Avalonia.Animation/KeyFramesStateMachine`1.cs:
```diff
@@ -0,0 +1,270 @@
+            _unsubscribe = true;
+            _currentState = KeyFramesStates.Disposed;
+        }
+        
```
ls

danwalmsley commented on src/Avalonia.Animation/KeyFrames`1.cs:
```diff
@@ -0,0 +1,179 @@
+    /// </summary>
+    public abstract class KeyFrames<T> : AvaloniaList<KeyFrame>, IKeyFrames
+    {
+        
```
ls

danwalmsley commented on src/Avalonia.Animation/KeyFrames`1.cs:
```diff
@@ -0,0 +1,179 @@
+
+            foreach (KeyFrame k in this)
+            {
+
```
ls

danwalmsley commented on src/Avalonia.Animation/RepeatBehavior.cs:
```diff
@@ -0,0 +1,13 @@
+
+namespace Avalonia.Animation
+{
+    public enum RepeatBehavior
```
add xml documentation.

danwalmsley commented on src/Avalonia.Animation/PlayState.cs:
```diff
@@ -0,0 +1,13 @@
+
+namespace Avalonia.Animation
+{
+    public enum PlayState
```
add xml docs

danwalmsley commented on src/Avalonia.Animation/Transition`1.cs:
```diff
@@ -0,0 +1,71 @@
+        }
+
+    }
+
```
ls

danwalmsley commented on src/Avalonia.Animation/Utils/BounceEaseUtils.cs:
```diff
@@ -0,0 +1,39 @@
+                return (54d / 5.0d * p * p) - (513d / 25.0d * p) + 268d / 25.0d;
+            }
+        }
+ 
```
ls

danwalmsley commented on src/Avalonia.Controls/ProgressBar.cs:
```diff
@@ -105,21 +105,21 @@ private void ValueChanged(AvaloniaPropertyChangedEventArgs e)
             private WeakReference<ProgressBar> _progressBar;
-            private IDisposable _indeterminateBindSubscription;
-            private TimeSpan _startTime;
+            //private IDisposable _indeterminateBindSubscription;
```
Is it possible to do something with this before merge, its preferable not to have commented code in master.

danwalmsley commented on src/Avalonia.Styling/Styling/Style.cs:
```diff
@@ -101,6 +113,21 @@ public void Attach(IStyleable control, IStyleHost container)
+
+                        var sub = animation.Apply((Animatable)control, obsMatch);
+                        subs.Add(sub);
+
```
ls

danwalmsley commented on src/Avalonia.Visuals/Animation/CrossFade.cs:
```diff
@@ -58,30 +58,33 @@ public CrossFade(TimeSpan duration)
-                    0,
-                    LinearEasing.For<double>(),
-                    Duration).ToTask());
+                //tasks.Add(Animate.Property(
```
commented code?

danwalmsley commented on src/Avalonia.Visuals/Media/Color.cs:
```diff
@@ -80,7 +81,7 @@ public static Color FromUInt32(uint value)
             );
         }
-
+        
```
looks like this file doesnt have any changes, can we revert it back?

danwalmsley commented on src/Markup/Avalonia.Markup.Xaml/PortableXaml/AvaloniaDefaultTypeConverters.cs:
```diff
@@ -12,6 +12,8 @@
 using Avalonia.Collections;
 using Avalonia.Controls.Templates;
+using Avalonia.Animation;
+using Avalonia.Animation.Easings;
```
Can we revert this change, seems not needed?

danwalmsley reviewed (dismissed):
@jmacato respect for this great effort, mostly iv just found code style nits (please don't get too annoyed with me, its just to try and help keep the code base consistent)

So far looking good though.

Iv played around with it quite a bit and its really powerful :)

grokys commented:
This is excellent work! I tried to write down my initial thoughts on giving it a spin below. This isn't really intended to be a review, more a place to maybe explore things from.

# Problems

## Exception Thrown

This animation:

```
<Animation Duration="0:0:0.5">
  <TransformKeyFrames Property="ScaleTransform.ScaleX">
    <KeyFrame Cue="0%" Value="0.1"/>
    <KeyFrame Cue="100%" Value="1"/>
  </TransformKeyFrames>
  <TransformKeyFrames Property="ScaleTransform.ScaleY">
    <KeyFrame Cue="0%" Value="0.1"/>
    <KeyFrame Cue="100%" Value="1"/>
  </TransformKeyFrames>
  <TransformKeyFrames Property="RotateTransform.Angle">
    <KeyFrame Cue="0%" Value="30"/>
    <KeyFrame Cue="50%" Value="-10"/>
    <KeyFrame Cue="70%" Value="3"/>
  </TransformKeyFrames>
  <DoubleKeyFrames Property="Opacity">
    <KeyFrame Cue="0%" Value="0"/>
    <KeyFrame Cue="100%" Value="1"/>
  </DoubleKeyFrames>
</Animation>
```

Throws:

```
System.InvalidOperationException
  HResult=0x80131509
  Message=Sequence contains no elements
  Source=System.Core
  StackTrace:
   at System.Linq.Enumerable.First[TSource](IEnumerable`1 source)
   at Avalonia.Animation.KeyFrames`1.GetKFPairAndIntraKFTime(Double t) in D:\projects\AvaloniaUI\Avalonia\src\Avalonia.Animation\KeyFrames`1.cs:line 75
```

# Design Thoughts

## Error: RepeatCount should be defined when RepeatBehavior is set to Repeat

Maybe should just repeat infinitely if RepeatCount is not set?

## Controls without RenderTransforms

If I apply a style with a transform animation to a control without
a `RenderTransform` I get an error:

```
Cannot apply keyframe: property ScaleX is not a transform.
```

This isn't ideal. It should either do nothing, or even better, add the missing transform, as manually adding transforms to each control is a PITA.

What to do if there's already a transform? Nothing? Create a TransformGroup? Not sure

## Having to copy/paste keyframes for ScaleX/Y

Is a pain:

```
<Animation Duration="0:0:0.5" FillMode="Forward">
  <TransformKeyFrames Property="ScaleTransform.ScaleX">
    <KeyFrame Cue="0%" Value="0.1"/>
    <KeyFrame Cue="60%" Value="1.2"/>
    <KeyFrame Cue="100%" Value="1"/>
  </TransformKeyFrames>
  <TransformKeyFrames Property="ScaleTransform.ScaleY">
    <KeyFrame Cue="0%" Value="0.1"/>
    <KeyFrame Cue="60%" Value="1.2"/>
    <KeyFrame Cue="100%" Value="1"/>
  </TransformKeyFrames>
</Animation>
```

In CSS you can just do:

```
@keyframes bounceIn {
  0% {
    transform: scale(0.1);
    opacity: 0;
  }
  60% {
    transform: scale(1.2);
    opacity: 1;
  }
  100% {
    transform: scale(1);
  }
}
```

We could perhaps add a `Scale` property to `ScaleTransform` to fix this?

## Syntax is a bit long-winded generally

Following on from the above, I Tried converting 
[this](https://github.com/daneden/animate.css/blob/da9afc06e8aa563baae6f406a0a844999ccc066b/animate.css#L2757)
animation for use in avalonia:

```
@keyframes jackInTheBox {
  from {
    opacity: 0;
    transform: scale(0.1) rotate(30deg);
  }

  50% {
    transform: rotate(-10deg);
  }

  70% {
    transform: rotate(3deg);
  }

  to {
    opacity: 1;
    transform: scale(1);
  }
}
```

The animation in avalonia is quite a bit less readable:

```
<Animation Duration="0:0:0.5">
  <TransformKeyFrames Property="ScaleTransform.ScaleX">
    <KeyFrame Cue="0%" Value="0.1"/>
    <KeyFrame Cue="100%" Value="1"/>
  </TransformKeyFrames>
  <TransformKeyFrames Property="ScaleTransform.ScaleY">
    <KeyFrame Cue="0%" Value="0.1"/>
    <KeyFrame Cue="100%" Value="1"/>
  </TransformKeyFrames>
  <TransformKeyFrames Property="RotateTransform.Angle">
    <KeyFrame Cue="0%" Value="30"/>
    <KeyFrame Cue="50%" Value="-10"/>
    <KeyFrame Cue="70%" Value="3"/>
  </TransformKeyFrames>
  <DoubleKeyFrames Property="Opacity">
    <KeyFrame Cue="0%" Value="0"/>
    <KeyFrame Cue="100%" Value="1"/>
  </DoubleKeyFrames>
</Animation>
```

Would something like the following be possible (or even desirable)?

```
<Animation Duration="0:0:0.5">
  <KeyFrame Cue="0%">
    <Setter Property="Opacity" Value="0"/>
    <Setter Property="ScaleTransform.Scale" Value="0.1"/>
    <Setter Property="RotateTransform.Angle" Value="30"/>
  </KeyFrame>
  <KeyFrame Cue="50%">
    <Setter Property="RotateTransform.Angle" Value="-10"/>
  </KeyFrame>
  <KeyFrame Cue="70%">
    <Setter Property="RotateTransform.Angle" Value="3"/>
  </KeyFrame>
  <KeyFrame Cue="100%">
    <Setter Property="Opacity" Value="1"/>
    <Setter Property="ScaleTransform.Scale" Value="1"/>
  </KeyFrame>
</Animation>
```

In reality, it's only 2 lines shorter, but perhaps more readable because it's grouped by keyframe rather than by property which is perhaps a more natural way of thinking about an animation? Just thinking out loud by this point really. I guess the syntax we have would be more familiar to WPF/UWP users.

grokys commented on src/Avalonia.Animation/Easing/Easing.cs:
```diff
@@ -0,0 +1,59 @@
+            {
+                _easingTypes = new Dictionary<string, Type>();
+
+                var derivedTypes = AppDomain.CurrentDomain.GetAssemblies()
```
Does this need to be extensible? As @danwalmsley notes below, the classes all need to be in the same namespace, so maybe not? If not, this should just be a hardcoded list or something. Or we might want to hardcode it now and make it extensible somehow later?

grokys commented:
Though looking at that final point again, I notice that the version with a single `Scale` property instead of `ScaleX` and `ScaleY` would be the most concise syntax. Maybe that'd be best. The downsides remain though that it's grouping by property instead of by keyframe and that you have to select the right type of keyframe. Maybe the first is only a problem when converting an animation from CSS?

danwalmsley commented:
That second example could look like this though...
```
<Animation Duration="0:0:0.5">
  <KeyFrame Cue="0%">
    <TransformKeyFrames Property="ScaleTransform.ScaleX" Value="1">
    <DoubleKeyFrames Property="Opacity" Value="0"/>
 </KeyFrame>
   <KeyFrame Cue="100%">
    <TransformKeyFrames Property="ScaleTransform.ScaleX" Value="0.5">
    <DoubleKeyFrames Property="Opacity" Value="1"/>
 </KeyFrame>
</Animation>
```

then I also wonder if we can do something like

```
<Animation Duration="0:0:0.5">
  <KeyFrame Cue="0%">
    <KeyFrameSetter Property="ScaleTransform.ScaleX" Value="1">
    <KeyFrameSetter Property="Opacity" Value="0"/>
 </KeyFrame>
   <KeyFrame Cue="100%">
    <KeyFrameSetter Property="ScaleTransform.ScaleX" Value="0.5">
    <KeyFrameSetter Property="Opacity" Value="1"/>
 </KeyFrame>
</Animation>
```

where `KeyFrameSetter` when xaml creates the type it somehow is actually given a derived class that fits the property type best.

grokys commented:
@danwalmsley the example I was talking about was (in CSS):

```
@keyframes jackInTheBox {
  from {
    opacity: 0;
    transform: scale(0.1) rotate(30deg);
  }

  50% {
    transform: rotate(-10deg);
  }

  70% {
    transform: rotate(3deg);
  }

  to {
    opacity: 1;
    transform: scale(1);
  }
}
```

danwalmsley commented:
oh yeah what iv written is a bit weird and almost the same as the example you posted.

```
<Animation Duration="0:0:0.5">
  <KeyFrame Cue="0%">
    <Setter Property="Opacity" Value="0"/>
    <Setter Property="ScaleTransform.Scale" Value="0.1"/>
    <Setter Property="RotateTransform.Angle" Value="30"/>
  </KeyFrame>
  <KeyFrame Cue="50%">
    <Setter Property="RotateTransform.Angle" Value="-10"/>
  </KeyFrame>
  <KeyFrame Cue="70%">
    <Setter Property="RotateTransform.Angle" Value="3"/>
  </KeyFrame>
  <KeyFrame Cue="100%">
    <Setter Property="Opacity" Value="1"/>
    <Setter Property="ScaleTransform.Scale" Value="1"/>
  </KeyFrame>
</Animation>
```

Perhaps KeyFrame we could have some helper classes that give you shortcuts for the transform ones like this:

```
  <KeyFrame Cue="0%">
    <Setter Property="Opacity" Value="0"/>
    <Scale Value="0.1"/>
    <Rotate Value="30"/>
  </KeyFrame>
  <KeyFrame Cue="50%">
    <Rotate>-10</Rotate>
  </KeyFrame>
  <KeyFrame Cue="70%">
    <Rotate Value="3"/>
  </KeyFrame>
  <KeyFrame Cue="100%">
    <Setter Property="Opacity" Value="1"/>
    <Scale Value="1"/>
  </KeyFrame>
</Animation>
```

grokys commented:
Yeah the problem with that is, in a word (well acronym) XAML. In XAML those `<Scale>` and `<Rotate>` elements have to be classes, which means that they'd be global to our namespace, meaning that we couldn't use them elsewhere. Maybe not a problem but I'd be hesitant to reserve such global concepts as `Scale` and `Rotate` to this syntax,

grokys commented:
Maybe even:

```
<KeyFrame Cue="0%" Content="Scale=0.1,Rotate=30,Opacity=1)" />
```

This could perhaps be built on top of the current syntax? Only problem here would be intellisense.

danwalmsley commented:
I cant get my head around this, but can a markup extension be used so that intellisense is maintained, and provide similar syntax, see for example whats been done with x:Bind...

https://docs.microsoft.com/en-us/windows/uwp/xaml-platform/x-bind-markup-extension

Commit:
Fix extraneous empty lines.
Fix Animation.cs braces

Commit:
Nit fixes part 2

Commit:
Fix CI

jmacato commented:
Should i add uniform properties to the transforms now (`Scale` instead of `ScaleX` & `ScaleY`, etc) ?
It wouldn't be a problem to rework the syntax to that but I'll need input on the implementation details later on

grokys commented:
Let's worry about the `Scale`, `ScaleX`, `ScaleY` stuff later I think, it shouldn't change anything here I don't think.

Commit:
Implement KeyFrame neutral values.

Commit:
Enable smarter TransformKeyFrames logic

Commit:
Fix introduced bug in neutral keyframe values.

Commit:
Fix all remaining nits - this commit will be used as a breakpoint for the upcoming syntax reimplementation.

jmacato committed:
Merge branch 'master' into wip-animations

Commit:
Start of syntax rework

Commit:
More syntax rework

Commit:
Merge branch 'wip-animations' of https://github.com/jmacato/Avalonia into wip-animations

Commit:
Finish rough implementation of the new syntax.

jmacato committed:
Merge branch 'master' into wip-animations

Commit:
Port RenderTest animations to the new syntax

Commit:
Simplify RepeatBehavior into a RepeatCount with Loop option.

Commit:
Fix nits

Commit:
Fix nits

Commit:
Merge branch 'wip-animations' of https://github.com/jmacato/Avalonia into wip-animations

jmacato committed:
Merge branch 'master' into wip-animations

grokys commented:
Looks great! Once #1594 has been merged we should be able to use the standard `<Setter>` instead of `<DoubleSetter>`, but I don't want to hold this up waiting for that!

Commit:
Merge Conflict fix

Commit:
Merge branch 'master' of https://github.com/AvaloniaUI/Avalonia into wip-animations
